docs/proposals/env-vars.md

```diff
@@ -0,0 +1,563 @@
+- `DisallowedEnvironmentVariables` in the `BuildStrategySpec` is considered first.  Any use of those keys in 
+  `BuildStep`, `BuildStrategySpec.[]EnvironmentVariable`, `BuildSpec.[]EnvironmentVariable`, `BuildRunSpec.[]EnvironmentVariable` results in a validation error.
+- `BuildStep` from the `BuildStrategySpec` is next.  The key/value pair from these override anything from
+  `BuildStrategySpec.[]EnvironmentVariable`, `BuildSpec.[]EnvironmentVariable`, `BuildRunSpec.[]EnvironmentVariable`
+- `BuildRunSpec.[]EnvironmentVariable` is next
+- followed by `BuildSpec.[]EnvironmentVariable`
+- and lastly, `BuildStrategySpec.[]EnvironmentVariable`
+
+Or conversely, the environment variables are applied in this fashion, cross-referencing with `BuildStrategySpec.AllowedEnvironmentVariables` to see if they are allowed.
+- first from `BuildStrategySpec.[]EnvironmentVariable` assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariables` are not used
+- then from `BuildSpec.[]EnvironmentVariable`, assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariables` are not used,
+  where values are replaced from any entries previously seeded
+- then from `BuildRunSpec.[]EnvironmentVariable`, assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariable` are not used,
+  where values are replaced from any entries previously seeded
+- then from `BuildStrategySpec.[]BuildStep`, assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariable` are not used,
+  where values are replaced from any entries previously seeded  
+
+To reiterate a key detail, the above methodology also provide a means of declaring whether an environment variable key/value setting can and connect be overridden. 
+If you want to ensure that environment variable key/value settings are *NOT* overridden, specify them in the `BuildStep.`  If you want
+to allow them to be overridden, specify them in the `BuildStrategySpec.[]EnvironmentVariable`, `BuildStrategySpec.[]EnvironmentVariable`, `BuildSpec.[]EnvironmentVariable`, or `BuildRunSpec.[]EnvironmentVariable`
+
+Lastly, some yaml to complete the visualization.
+
+`BuildStrategy` with a disallow entry, a spec level default, and a final, always set via a `BuildStep`:
+
+```yaml
+apiVersion: shipwright.io/v1alpha1
+kind: ClusterBuildStrategy
+metadata:
+  name: something-that-uses-docker
+spec:
+  # nobody is allowed to change DOCKER_TLS_VERIFY
+  disallowedEnvironmentVariables:
+    - 'DOCKER_TLS_VERIFY'
+  environmentVariables:
+    # for convenience, we add the version most our users can use, but we'll allow you to override
+    - name: DOCKER_API_VERSION
+      value: '1.19'
+  buildSteps:
+    - command:
+        - /usr/local/bin/docker
+        - build
+      env:
+        - name: DOCKER_CONFIG
+          value: /tekton/home/.docker
```

Comment:
I am a bit confused by this API. Are `environmentVariables` applied to every build step with the value provided? Can a developer override the `DOCKER_CONFIG` env var?

I would almost rather have build strategy authors define the env vars in line with the build steps, and not worry about (redundantly) declaring default values in `environmentVariables`. Values that can't be overrode should be declared clearly and not rely on hidden behavior. As I understand the behavior in the EP, I think the following would provide more clarity:

```suggestion
  disallowedEnvironmentVariables:
    # Don't allow docker to push to an insecure registry
    - 'DOCKER_TLS_VERIFY'
    # Don't allow the pull/push secrets to be overrode from what Shipwright provides
    - 'DOCKER_CONFIG'
  buildSteps:
    - command:
        - /usr/local/bin/docker
        - build
      env:
        - name: DOCKER_CONFIG
          value: /tekton/home/.docker
            environmentVariables:
       # for convenience, we add the version most our users can use, but developers can override at build time.
       - name: DOCKER_API_VERSION
         value: '1.19'
```

Reply:
Yes they are applied to build steps, but then those k/v paris could be overwritten by the same keys specified in builds or build runs.

Where as the items specified directly in the BuildStep cannot be overridden.

But assuming we go with this "administrative control", let's dive into the alternative noted by @SaschaSchwarze0 in https://github.com/shipwright-io/build/pull/726#issuecomment-833824197

```
steps:
  - name: my-build-tool
    env:
      - name: MY_FIXED_ENVVAR
	    value: "fixed value"
	    overwritable: false             # default: true
      - name: MY_DEFAULT_VALUE_ENVVAR
	    value: "default value"
```

Well, if you consider the current def of BuildStep:

```
type BuildStep struct {
	corev1.Container `json:",inline"`
}
```

To achieve that yaml, we need to either 

a) create a `shipwright.Container` type, with all the existing `corev1.Container` fields, but with the `envVar` array field change to have to be an array field of a new type that has the `overrwritable` flag.

That seems like a more disruptive change.  And fragile given the field complexity of `corev1.Container`.

or 

b) create a 

```
type BuildStep struct {
   corev1.Container `json:",inline"`
   EnvVarOverwritable   map[string]bool
}

```
Where the key to that map is the env var name, and then true/false indicates whether the build run or build can overwrite.

Of those two, at least for me b) is the more tolerable of the two.  But it is a slight departure from the yaml @SaschaSchwarze0 noted.

When I originally wrote this, b) was not as palatable as what is currently in the EP, but I could be talked into switching to b) now if the current form is too lacking in intuitiveness.